### ISSUE TITLE
Allow hyperbolic orbits to stay within the bounds of a plot

### ISSUE THREAD

**Issue by Juanlu001** [poliastro/poliastro#111]
After #19 was fixed plotting only one hyperbolic orbit looks okay, but there can be cases when one wants it to do something like this:

![new_horizons](https://cloud.githubusercontent.com/assets/316517/12539950/2ef2ac8e-c2fe-11e5-8d10-3a3067e6f3f1.png)

This plot needed a patch in `_generate_vals` (hence tagging it as a bug), but it only works for cases where there's a previous closed orbit plotted. The logic to handle both cases has to be more complex.

**Comment by Juanlu001** [comment]
I'm postponing this for the next release too.

**Comment by Juanlu001** [comment]
Asking the algorithm to `_adjust_to_bigger_bounded_orbit()` magically is probably too much. I should give more control to the user. Postponing.

**Comment by Juanlu001** [comment]
Blocked on #218.

**Comment by Juanlu001** [comment]
This is no longer blocked.

**Comment by sudnadja** [comment]
When plotting certain hyperbolic paths, the position plots of the objects appear incorrect. 
For example, in July of 2015, New Horizons flew past pluto. Taking position data from horizons for both Pluto and New Horizons with 500@0 origin and "frame" as the fundamental plane, those objects are placed into poliastro with: 

PlutoM = Orbit.from_vectors(Sun, [1.197648971483503E+09, -4.443903841678363E+09, -1.747653049536265E+09] * u.km , [5.382941983259522E+00,  8.185732351064959E-01, -1.389415786284875E+00] * u.km/u.s, Time('2015-07-14 07:59',scale='tdb') )
NewHorizonsM = Orbit.from_vectors(Sun, [1.197659243752796E+09, -4.443716685978071E+09, -1.747610548576734E+09] * u.km , [5.540549267188614E+00, -1.251544669134140E+01, -4.848892572767733E+00] * u.km/u.s, Time('2015-07-14 07:59',scale='tdb') )

New Horizons is nearly on top of Pluto at that point, however the plot places New Horizons much earlier on its path: 


![new horizons pluto encounter](https://user-images.githubusercontent.com/35163155/34637945-041e3858-f276-11e7-86ee-be2757538d0d.png)

And a notebook exhibiting this behavior: 

[New+Horizons+Pluto+Encounter.zip](https://github.com/poliastro/poliastro/files/1608594/New.Horizons.Pluto.Encounter.zip)

**Comment by Juanlu001** [comment]
I confirm I can reproduce this using both 0.8.0 and latest master version. Working on it

**Comment by Juanlu001** [comment]
With 0.7.0, it was also wrong but in a different way

![png image 754 x 712 pixels](https://user-images.githubusercontent.com/316517/34640093-9ce831d0-f2ec-11e7-9455-fe453b26e133.png)

**Comment by Juanlu001** [comment]
Forget about the old behavior, it's gone anyway.

Doing some quick experiments, it seems that the "first" point in the sampling is used as the "dot" where the spacecraft is. Specifically, these lines:

https://github.com/poliastro/poliastro/blob/v0.8.0/src/poliastro/plotting.py#L158-L159

Clearly, for hyperbolic orbits the assumption "the first point is the original `Orbit`" is wrong. The problem is somewhere here:

https://github.com/poliastro/poliastro/blob/v0.8.0/src/poliastro/twobody/orbit.py#L293-L320

However, I still don't understand why the elliptic case is managed correctly.

**Comment by Juanlu001** [comment]
The reason why is that sampling values for the elliptic case start at zero:

https://github.com/poliastro/poliastro/blob/44dbc6e26f2d0a6b395a910c39c9e8131c8946b2/src/poliastro/twobody/orbit.py#L294-L295

whereas for hyperbolas, start at the limit value:

https://github.com/poliastro/poliastro/blob/44dbc6e26f2d0a6b395a910c39c9e8131c8946b2/src/poliastro/twobody/orbit.py#L296-L300

The fix will need some refactoring, why may or may not address the original problem stated in https://github.com/poliastro/poliastro/issues/111#issue-128434697, which is that it should be easy to let hyperbolic orbits escape the bounds of a plot instead of redefining them.

**Comment by Juanlu001** [comment]
I created a new issue https://github.com/poliastro/poliastro/issues/295

**Comment by shreyasbapat** [comment]
Is this problem still there?

**Comment by Juanlu001** [comment]
Probably, because we did not change anything related to that :) But as usual I failed to provide code to reproduce the issue. Would you like to try? This plot comes from this notebook, which unfortunately is in Spanish:

http://nbviewer.jupyter.org/github/Pybonacci/notebooks/blob/master/Explorando%20el%20Planeta%20Nueve%20con%20Python%20usando%20poliastro.ipynb

**Comment by shreyasbapat** [comment]
Yeah Surely, I would try. I will just copy the code cells. And use google for Spanish!

**Comment by Juanlu001** [comment]
I stumbled upon this again while running the notebook I mentioned above. This is the code to reproduce with the current master:

```
ss_hyp = Orbit.from_vectors(
    Sun,
    [ -9.77441841e+07,  1.01000539e+08,  4.37584668e+07] * u.km,
    [ 23.75936985,-43.09599568, -8.7084724 ] * u.km / u.s,
)
ss_big = Orbit.from_vectors(
    Sun,
    [  4.62521605e+10,  1.55496163e+11, -4.18874038e+10] * u.km,
    [-0.46461615, 0.20385617, 0.24373322] * u.km / u.s
)
```

Plotting each of them separately works:

| ![hyp](https://user-images.githubusercontent.com/316517/36848594-ff2269ca-1d61-11e8-9ffd-449820885db4.png) | ![big](https://user-images.githubusercontent.com/316517/36848596-007ed1a0-1d62-11e8-9e55-85d6db56f074.png) |
|------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------|


(notice the slight bug in the hyperbolic one?)

However, plotting them together shows only the big, elliptic orbit (regardless of the order):

```
frame = OrbitPlotter()

frame.plot(ss_big)
frame.plot(ss_hyp)
```

![together](https://user-images.githubusercontent.com/316517/36848647-2ae0179c-1d62-11e8-9872-22c21b73abde.png)

Therefore, it's not enough to compute the limit `nu_values` for hyperbolic orbits inside the `sample` method: the plotting code should set them, depending on the bounds of the plot or some other parameter. Still not sure how to fix this.

**Comment by Juanlu001** [comment]
And by the way, this is the changes I had to made back in the day to properly plot this:

https://github.com/poliastro/poliastro/commit/9899f5a6f6f741e26ce0bae73f48eba569b3f5e8

```diff
diff --git a/src/poliastro/plotting.py b/src/poliastro/plotting.py
index c30f610..937801b 100644
--- a/src/poliastro/plotting.py
+++ b/src/poliastro/plotting.py
@@ -156,8 +156,8 @@ def _generate_vals(self, state):
 
         if state.ecc >= 1:
             # Select a sensible limiting value for non-closed orbits
-            # This corresponds to r = 3p
-            nu_limit = Angle(np.arccos(-(1 - 1 / 3.) / state.ecc))
+            # This corresponds to r = infinity
+            nu_limit = Angle(np.arccos(-1 / state.ecc))
             nu_invalid = ((nu_vals > nu_limit) &
                           (nu_vals < (-nu_limit).wrap_at('360d')))
             nu_vals[nu_invalid] = np.nan
```

**Comment by Juanlu001** [comment]
I'm blocking this issue until we have #274 working. That one is easier and will give us hints on how to fix this.

**Comment by Juanlu001** [comment]
This is not blocked anymore.

**Comment by shreyasbapat** [comment]
What exactly is the problem ? Is it like the hyperbolic plot goes till the edge of the plot or something else?

**Comment by Juanlu001** [comment]
the problem is that if you plot a "small" hyperbolic orbit and a big elliptic orbit, the hyperbolic one won't be shown, and if you plot a "big" hyperbolic orbit, it will push the boundaries of the plot and you won't be able to see anything. Not sure if this is an easily solvable problem, since in general plotting closed (elliptical) orbits of very different sizes will always showcase this problem as well. The current behavior treats hyperbolic orbits as closed orbits, and I wonder if we can do better.

**Comment by Juanlu001** [comment]
It's not actually clear what to do with this, and also the thread is confusing because we mixed the original problem with other bugs. For the moment, I'm removing the milestone.

**Comment by shreyasbapat** [comment]
What has to be done for this?

**Comment by Juanlu001** [comment]
Perhaps it's not really something we can solve. Our current APIs allow you to clip the axis and change its limits, so perhaps that's enough. I'm closing this.
